axlearn/cloud/gcp/jobset_utils.py

```diff
@@ -670,24 +670,220 @@ def __call__(self) -> Sequence[Nested[Any]]:
+        cfg: GPUReplicatedJob.Config = self.config
+        job_spec = dict(
+            spec=dict(
+                parallelism=cfg.accelerator.num_replicas,
+                completions=cfg.accelerator.num_replicas,
+                backoffLimit=0,  # Fail the job if any node fails. Retries happen at JobSet level.
+                template=self._build_pod(),
+            ),
+        )
+        # NOTE: the suffix here impacts how long job names can be.
+        return [dict(name="job", replicas=1, template=job_spec)]
+
+
+class A3UltraReplicatedJob(GPUReplicatedJob):
+    """Builds a replicated job spec for A3 Ultra GPU job, to be used with JobSet API."""
+
+    def _build_volumes(self) -> list[Any | dict[str, Nested]]:
+        return super()._build_volumes() + [
+            {
+                "name": "gib",
+                "hostPath": {"path": "/home/kubernetes/bin/gib"},
+            },
+        ]
+
+    def _build_main_container(self) -> Nested[Any]:
+        """Builds the config for the container running the job.
+
+        Returns:
+            A nested dict corresponding to a k8s Container config.
+        """
+        cfg: GPUReplicatedJob.Config = self.config
+
+        volume_mounts = [
+            {"name": "shared-memory", "mountPath": "/dev/shm"},
+            {"name": "nvidia-install-dir-host", "mountPath": "/usr/local/nvidia/lib64"},
+            {"name": "gib", "mountPath": "/usr/local/gib"},
+        ]
+
+        env_vars: dict[str, str] = {}
+        env_vars["DISTRIBUTED_COORDINATOR"] = f"{cfg.name}-job-0-0.{cfg.name}:8080"
+        env_vars["NUM_PROCESSES"] = f"{cfg.accelerator.num_replicas}"
+        env_vars["LD_LIBRARY_PATH"] = "/usr/local/nvidia/lib64"
+
+        default_xla_flags = [
+            # Maxtext XLA flags:
+            # https://github.com/AI-Hypercomputer/gpu-recipes/blob/dc6ef1afc1492f05e5741356f00cf645a9f1b795/src/helm-charts/a3ultra/maxtext-training/templates/maxtext-configmap.yaml#L26-L38
+            "--xla_gpu_enable_latency_hiding_scheduler=true",
+            "--xla_gpu_enable_triton_gemm=false",
+            "--xla_gpu_graph_level=0",
+            "--xla_gpu_all_reduce_combine_threshold_bytes=2147483648",
+            "--xla_gpu_all_gather_combine_threshold_bytes=2147483648",
+            "--xla_gpu_reduce_scatter_combine_threshold_bytes=16777216",
+            "--xla_gpu_enable_pipelined_all_gather=true",
+            "--xla_gpu_enable_pipelined_reduce_scatter=true",
+            "--xla_gpu_enable_pipelined_all_reduce=true",
+            "--xla_gpu_enable_all_gather_combine_by_dim=false",
+            "--xla_gpu_enable_reduce_scatter_combine_by_dim=false",
+            "--xla_disable_hlo_passes=rematerialization",
+            "--xla_gpu_enable_while_loop_double_buffering=true",
+        ]
+        env_vars["XLA_FLAGS"] = " ".join(default_xla_flags)
+
+        # NCCL flags needed
+        env_vars.update(
+            {
+                # Enable auto PGLE available in jax 0.4.33
```

Comment:
Does it still apply if we are already on jax 0.4.38

Reply:
removed this comment